packages/material-ui/src/Tooltip/Tooltip.js

```diff
@@ -72,119 +73,173 @@ export const styles = theme => ({
-      // not controlled, use internal state
-      this.state.open = false;
-    }
-  }
+  return mountedRef;
+}
 
-  componentDidMount() {
-    warning(
-      !this.childrenRef.disabled ||
-        (this.childrenRef.disabled && this.props.title === '') ||
-        this.childrenRef.tagName.toLowerCase() !== 'button',
-      [
-        'Material-UI: you are providing a disabled `button` child to the Tooltip component.',
-        'A disabled element does not fire events.',
-        "Tooltip needs to listen to the child element's events to display the title.",
-        '',
-        'Place a `div` container on top of the element.',
-      ].join('\n'),
-    );
+function Tooltip(props) {
+  const {
+    children,
+    classes,
+    disableFocusListener,
+    disableHoverListener,
+    disableTouchListener,
+    enterDelay,
+    enterTouchDelay,
+    id,
+    interactive,
+    leaveDelay,
+    leaveTouchDelay,
+    onClose,
+    onOpen,
+    open: openProp,
+    placement,
+    PopperProps,
+    theme,
+    title,
+    TransitionComponent,
+    TransitionProps,
+    ...other
+  } = props;
+
+  const [openState, setOpenState] = React.useState(false);
+  const [, forceUpdate] = React.useState(0);
+  const ignoreNonTouchEvents = React.useRef(false);
+  const { current: isControlled } = React.useRef(props.open != null);
+  const [childNode, setChildNode] = React.useState();
+  // can be removed once we drop support for non ref forwarding class components
+  const handleOwnRef = React.useCallback(ref => {
+    setChildNode(ReactDOM.findDOMNode(ref));
+  }, []);
+  const handleRef = useForkRef(childNode, handleOwnRef);
+  const defaultId = React.useRef();
+  const mountedRef = useMountedRef();
+  const closeTimer = React.useRef();
+  const enterTimer = React.useRef();
+  const leaveTimer = React.useRef();
+  const touchTimer = React.useRef();
+
+  React.useEffect(() => {
+    if (childNode) {
+      warning(
+        !childNode.disabled ||
+          (childNode.disabled && title === '') ||
+          childNode.tagName.toLowerCase() !== 'button',
+        [
+          'Material-UI: you are providing a disabled `button` child to the Tooltip component.',
+          'A disabled element does not fire events.',
+          "Tooltip needs to listen to the child element's events to display the title.",
+          '',
+          'Place a `div` container on top of the element.',
+        ].join('\n'),
+      );
+    }
+  }, [title, childNode]);
 
+  React.useEffect(() => {
     // Fallback to this default id when possible.
     // Use the random value for client side rendering only.
     // We can't use it server-side.
-    this.defaultId = `mui-tooltip-${Math.round(Math.random() * 1e5)}`;
+    defaultId.current = `mui-tooltip-${Math.round(Math.random() * 1e5)}`;
 
-    // Rerender with this.defaultId and this.childrenRef.
-    if (this.props.open) {
-      this.forceUpdate();
+    // Rerender with defaultId and childNode.
+    if (openProp && !mountedRef.current) {
```

Comment:
Do we need to take `mountedRef` into account?

```suggestion
    if (openProp) {
```

Reply:
I have tried this change live. The tests and the demos look OK. Looks like we can remove some code :)

Reply:
Haha, I thought the same thing